rfd/0185-k8s-access-non-cert-routing.md

```diff
@@ -0,0 +1,247 @@
+`KubernetesCluster` attribute will continue to function.
+
+- VX.Y.0:
+  - Introduce new Kube Forwarder behaviour.
+  - Introduce opt-in `kubernetes/v2` `tbot` output. This will allow
+    users to opt-in to the new behaviour once they are satisfied that their
+    Auth Server, Proxy and Kubernetes Agents have been upgraded to the required
+    version.
+- V(X+1).0.0:
+  - Switch `tsh` to new behaviour. No longer requests the signing of a
+    certificate with the `KubernetesCluster` attribute and uses standard user
+    identity to authenticate.
+  - Switch `tbot` `kubernetes/v1` output to use `kubernetes/v2` implementation
+    under the hood.
+- V(X+2).0.0:
+  - It is now "safe" to remove the old attribute based routing from the Kube
+    Forwarder. We may wish to delay this further if we have concerns about
+    older client compatability.
+
+#### Client Compatability
+
+Kubernetes API clients have been known to vary greatly in their support for 
+more unusual configuration. We have witnessed this in the past when it came to
+supporting the ability to explicitly provide the SNI to use when making
+requests.
+
+Therefore, we should ensure that Kubernetes clients and SDKs do support
+providing a base path as part of the server address (e.g they do not strip off
+the path and only use the hostname/port).
+
+The key customer we are working with has specified that they use the following
+Kubernetes clients:
+
+- `kubectl`
+- Python Kubernetes SDK
+
+Both of these clients correctly handle a base-path as configured in Kubeconfig:
+
+```yaml
+clusters:
+  - cluster:
+      server: http://127.0.0.1:8888/example/base/path
+    name: test-base-path
+```
+
+#### Analytics
+
+It would no longer be possible to track usage of Kubernetes Access for analytics
+purposes via the certificate generation event. However, we have other events
+that indicate Kubernetes Access usage (e.g an event per request), that are
+better suited for this purpose - so this seems acceptable.
+
+#### TBot UX
+
+The `tbot` configuration will be extended with a new `kubernetes/v2` service
+type that will allow you to specify multiple kubernetes clusters to generate
+within a single Kubeconfig file.
+
+To select which clusters to include within the Kubeconfig, the name or labels
+of the clusters can be specified.
+
+```yaml
+- type: kubernetes/v2
+  destination:
+    type: directory
+    path: /opt/machine-id
+  # Each selector can either be a name or a set of labels. If multiple selectors
+  # are provided, then the output will contain the union of the results of the
+  # selectors.
+  selectors:
+    - name: my-cluster
+    - labels:
+        "env": "production"
+```
+
+### Alternatives
+
+#### Leverage SNI
+
+Rather than encoding the target Kubernetes cluster into the URL, we could
+instead encode it into the SNI.
+
+This has the following limitations:
+
+- We've already witnessed limited support for specifying SNIs in Kubernetes
```

Comment:
The main challenge is that by using SNI for routing and supporting `kubectl`, Teleport would need to generate certificates for `*.rootCluster.kube.teleportproxy.com`, as well as for leaf clusters `*.leafCluster.kube.teleportproxy.com`. Over time, this approach could introduce significant issues as the system evolves. Since we must support `kubectl`, the only viable solution is to send the identification via `tls-server-name`.